clang/include/clang/CIR/Interfaces/ASTAttrInterfaces.td

```diff
@@ -0,0 +1,213 @@
+//===- ASTAttrInterfaces.td - CIR AST Interface Definitions -----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_CIR_INTERFACES_AST_ATTR_INTERFACES
+#define MLIR_CIR_INTERFACES_AST_ATTR_INTERFACES
+
+include "mlir/IR/OpBase.td"
+
+let cppNamespace = "::mlir::cir" in {
+
+    def ASTDeclInterface : AttrInterface< "ASTDeclInterface" > {
+        let methods = [
+            InterfaceMethod< "", "bool", "hasOwnerAttr", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->template hasAttr< clang::OwnerAttr >();
+                }]
+            >,
+            InterfaceMethod< "", "bool", "hasPointerAttr", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->template hasAttr< clang::PointerAttr >();
+                }]
+            >,
+            InterfaceMethod< "", "bool", "hasInitPriorityAttr", (ins), [{}],
+                /*defaultImplementation=*/ [{
+                    return $_attr.getAstDecl()->template hasAttr< clang::InitPriorityAttr >();
+                }]
+            >
+        ];
+    }
+
+    def ASTDeclContextInterface : AttrInterface< "ASTDeclContextInterface" > {
+        let methods = [
+            InterfaceMethod< "",
+                "llvm::SmallVector< ::mlir::Attribute, 4 >", "decls",
```

Comment:
I'm a bit skeptical that we'll wanna use this method in practice:
- Walking all decls() just to wrap them can get a bit expensive. An `ArrayRef` on ClangAST is much more convenient, but I understand why you cannot do it here.
- Do we want it to work recursively? Here you map all direct attached decls, but what does that mean for decls of decls?

I'm assuming you added this to handle the case where we detect `promise_type` - is there a way we could instead expose a method `isPromiseType` that allows each AST provider to compute that information instead? Perhaps that could also be useful for other future things (maybe?)

Not a deal breaker, just curious about the design space. (For more background: I might change the lifetime checker to detect such types by looking into the CIR types instead, and we could later remove this)

Reply:
My first try was to create `ASTDeclIteratorInterface` but it turned out to be too complex. I like the idea to just provide `isPromiseType` which will be easier for us to implement too :)

Reply:
implemented in 151eb61c93680204468e921b8eead46d2037238f